Add learn page on `Building a Data Service` (#3469)
ballerina-platform/ballerina-dev-website: learn/getting-started/building-a-data-service/building-a-data-service.md

---
layout: ballerina-building-a-data-service-left-nav-pages-swanlake
title: Building a Data Service in Ballerina 
description: Connecting to a MySQL database and executing queries using an HTTP RESTful API using Ballerina.
keywords: ballerina, data service, mysql, database, REST, API
permalink: /learn/building-a-data-service/
active: building-a-data-service
intro: Connecting to a MySQL database and executing queries using an HTTP RESTful API using Ballerina.
redirect_from:
 - /learn/building-a-data-service
---

This tutorial will cover how you can connect to a MySQL database and perform queries against it using Ballerina. 
In addition, this tutorial will also elaborate on how you can create an HTTP RESTful API using Ballerina that can be used

> by end-users to perform basic CRUD operations on the database.

Reviewer: ```suggestion
to perform basic CRUD operations on the database.
```

> ![Data Service Architecture](/learn/images/data-service-architecture.png "Data Service Architecture")

Author: Image not yet added

The methodology outlined in this tutorial can be used to work with PostgreSQL, SQL Server, OracleDB or any other 
relational database as well using the [`PostgreSQL`](https://central.ballerina.io/ballerinax/postgresql),
[`MSSQL`](https://central.ballerina.io/ballerinax/mssql), [`OracleDB`](https://central.ballerina.io/ballerinax/oraceldb)
and [`JDBC`](https://central.ballerina.io/ballerinax/java.jdbc) connectors for Ballerina respectively.

## Prerequisites

### Setting up a MySQL Server Instance

There are several methods you can use to set up a MySQL server.

> 1. Install a MySQL server on your machine locally by downloading and installing MySQL as detailed on the [MySQL website](https://dev.mysql.com/doc/mysql-getting-started/en/#mysql-getting-started-installing) for different platforms.

Reviewer: ```suggestion
1. Install a MySQL server on your machine locally by downloading and installing [MySQL](https://dev.mysql.com/doc/mysql-getting-started/en/#mysql-getting-started-installing) for different platforms.
```

> 2. Using a cross-platform web-server solution such as [XAMPP](https://www.apachefriends.org/index.html) or [WampServer](https://www.wampserver.com/en/).

Reviewer: ```suggestion
2. Using a cross-platform web-server solution such as [XAMPP](https://www.apachefriends.org/index.html) or [WampServer](https://www.wampserver.com/en/).
```
using the `ing` form here is inconsistent with the other points.

3. Using [Docker](https://dev.mysql.com/doc/mysql-installation-excerpt/8.0/en/docker-mysql-getting-started.html) to create a MySQL server deployment.
4. Using a cloud-based MySQL solution such as Google’s [CloudSQL](https://cloud.google.com/sql), Amazon’s [RDS for MySQL](https://aws.amazon.com/rds/sqlserver/), or Microsoft’s [Azure Database for MySQL](https://azure.microsoft.com/en-us/services/mysql/).

> >Be sure to note down the connection and authentication details for connecting to the MySQL server including the hostname, port, username and password.

Reviewer: ```suggestion
>**Note:** Keep the connection and authentication details seven for connecting to the MySQL server including the hostname, port, username, and password.
```

### Creating a Database and Table

Connect to the MySQL server using the terminal (or any other preferred method) and execute the following commands to 

> create a database and table. This tutorial will demonstrate the basic use-case of creating, maintaining and

Reviewer: ```suggestion
create a database and table. This tutorial will demonstrate the basic use-case of creating, maintaining and
```
Check the Oxford comma in all occurrences.

interacting with a database of employees in an organization.

```roomsql
CREATE DATABASE IF NOT EXISTS Company;

CREATE TABLE Company.Employees (
	employee_id INTEGER AUTO_INCREMENT PRIMARY KEY,
	first_name  VARCHAR(255) NOT NULL,
	last_name   VARCHAR(255) NOT NULL,
	email       VARCHAR(255) NOT NULL,
	phone       VARCHAR(50) NOT NULL ,
	hire_date   DATE NOT NULL,
	manager_id  INTEGER REFERENCES Employees(employee_id),
	job_title   VARCHAR(255) NOT NULL
)
```

### Creating a Ballerina Project

A new Ballerina project can be created by executing the following command in the directory in which you want to create the project.

```shell
bal new data_service
``` 

> >For more information on creating Ballerina packages, see [Creating your First Ballerina Package](/learn/hello-world/creating-your-first-ballerina-package/#creating-the-package/).

Reviewer: ```suggestion
>For more information on creating Ballerina packages, see [Creating your First Ballerina Package](/learn/hello-world/creating-your-first-ballerina-package/#creating-the-package/).
```
Please correct the format of the info, tips, and note sections in all occurrences.

### Creating a Record to Represent an Employee

In Ballerina, records are a data-type that maps keys to values. You can define a closed record to represent a single row
in the `Employees` table in the `main.bal` file.

```ballerina
public type Employee record {|
    int employee_id?;
    string first_name;
    string last_name;
    string email;
    string phone;
    time:Date hire_date;
    int? manager_id;
    string job_title;
|};
```

This record-type will be the basis for interacting with the database.

## Connecting to and Interacting with the Database

### Adding the MySQL Driver

> The MySQL driver JAR is necessary in order to connect to and interact with a MySQL server. There are several methods of doing this.

Reviewer: ```suggestion
The MySQL driver JAR is necessary to connect to and interact with a MySQL server. There are several methods of doing this.
```

configurable string USER = ?;
configurable string PASSWORD = ?;
configurable string HOST = ?;
configurable int PORT = ?;
configurable string DATABASE = ?;
```

>For more information on defining configurable variables in Ballerina, see [Defining Configurable Variables](/learn/making-ballerina-programs-configurable/defining-configurable-variables/).

### Connecting to the Database

Firstly, the [`MySQL`](https://central.ballerina.io/ballerinax/mysql) 
and [`SQL`](https://central.ballerina.io/ballerina/sql) packages must be imported into your `main.bal` file.

```ballerina
import ballerinax/mysql;
```

The `mysql:Client` can be used to connect to the database. Include the following code in your `main.bal` file and
execute the program using `bal run`. If the program runs without throwing an error, 
that indicates that the connection has been established successfully. This client can be defined globally and be
used across all parts of the program.

```ballerina
final mysql:Client dbClient = check new(
    host=HOST, user=USER, password=PASSWORD, port=PORT, database="Company"
);
```

>The MySQL package provides additional connection options and the ability to configure connection pool 

> >properties when connecting to the database which is not covered in this tutorial. To learn more about this, 

Reviewer: ```suggestion
>properties when connecting to the database, which are not covered in this tutorial. To learn more about this, 
```

> >see the API docs for [`mysql:Client`](https://lib.ballerina.io/ballerinax/mysql/1.2.0/clients/Client).

Reviewer: ```suggestion
>see [`mysql:Client`](https://lib.ballerina.io/ballerinax/mysql/1.2.0/clients/Client).
```
Please check the guidelines on giving reference links.

### Executing Queries

> `mysql:Client` provides two primary remote methods for performing queries.

Reviewer: ```suggestion
The `mysql:Client` provides two primary remote methods for performing queries.
```

1. `query()` - Executes an SQL query and returns the results (rows) from the query. 

>    `queryRow()` is a variation of this method, which returns at most, a single row from the result.

Reviewer: ```suggestion
   The `queryRow()` is a variation of this method, which returns at most a single row from the result.
```

        `SELECT * FROM Employees`
    );
    check from Employee employee in resultStream
        do {
            employees.push(employee);
        };
    check resultStream.close();
    return employees;
}

isolated function updateEmployee(Employee emp) returns int|error {
    sql:ExecutionResult result = check dbClient->execute(`
        UPDATE Employees SET
            first_name = ${emp.first_name}, 
            last_name = ${emp.last_name},
            email = ${emp.email},
            phone = ${emp.phone},
            hire_date = ${emp.hire_date}, 
            manager_id = ${emp.manager_id},
            job_title = ${emp.job_title}
        WHERE employee_id = ${emp.employee_id}  
    `);
    int|string? lastInsertId = result.lastInsertId;
    if lastInsertId is int {
        return lastInsertId;
    } else {
        return error("Unable to obtain last insert ID");
    }
}

isolated function removeEmployee(int id) returns int|error {
    sql:ExecutionResult result = check dbClient->execute(`
        DELETE FROM Employees WHERE employee_id = ${id}
    `);
    return result.affectedRowCount;
}
```

## Exposing the Database via an HTTP RESTful API

> After you have defined the methods necessary to manipulate the database, you can expose these selectively via an HTTP

Reviewer: ```suggestion
After you have defined the methods necessary to manipulate the database, expose these selectively via an HTTP
```

> RESTful API. For this, you would first need to import the Ballerina [`HTTP` module](https://central.ballerina.io/ballerina/http).

Reviewer: ```suggestion
RESTful API. For this, import the Ballerina [`HTTP` module](https://central.ballerina.io/ballerina/http).
```
Check the guidelines on addressing the direct active voice.

Reviewer: Please check and fix all occurrences.

```ballerina
import ballerina/http;
```

### Creating a Service

> Afterwards, you can create a service as follows. This would create an endpoint `/employees` on port `8080` which can 

Reviewer: ```suggestion
Afterwards, you can create a service as follows. This would create an `/employees` endpoint 
 on port `8080`, which can 
```

```ballerina
service /employees on new http:Listener(8080) {
}
```

### Creating Resource Functions

Within this service, you can define resource functions to provide access to the database. The code snippet below 
demonstrates a resource function that can be used to create a new employee via a `POST` request.

```ballerina
service /employees on new http:Listener(8080) {

    	isolated resource function post .(@http:Payload Employee emp) returns int|error? {
            return addEmployee(emp);
        }
        
}
```

This resource function may be invoked by sending a `POST` request to `http://localhost:8080/employees` with the 
required data as a JSON payload.

```shell
curl --location --request POST 'http://localhost:8080/employees/' \
    --header 'Content-Type: text/plain' \
    --data-raw '{
        "employee_id": 6,
        "first_name": "test",
        "last_name": "test",
        "email": "test@test.com",
        "phone": "882 771 110",
        "hire_date": {
            "year": 2021,
            "month": 12,
            "day": 16
        },
        "manager_id": 1,
        "job_title": "Sales Manager"
    }'
```

> Similarly, you can define resource functions within the service for each of our use cases as demonstrated below.

Reviewer: ```suggestion
Similarly, you can define resource functions within the service for each of your use cases as demonstrated below.
```
Please fix all occurrences.

```ballerina
service /employees on new http:Listener(8080) {

    isolated resource function post .(@http:Payload Employee emp) returns int|error? {
        return addEmployee(emp);
    }
    
    isolated resource function get [int id]() returns Employee|error? {
        return getEmployee(id);
    }
    
    isolated resource function get .() returns Employee[]|error? {
        return getAllEmployees();
    }
    
    isolated resource function put .(@http:Payload Employee emp) returns int|error? {
        return updateEmployee(emp);
    }
    
    isolated resource function delete [int id]() returns int|error? {
        return removeEmployee(id);       
    }

}
```

>**Note:** 

> >This tutorial does not cover topics such as authentication, rate-limiting, usage monitoring, billing, caching, 

Reviewer: ```suggestion
>This tutorial does not cover topics such as authentication, rate-limiting, usage monitoring, billing, caching, 
```
Do we need to mention this explicitly?

Author: @daneshk thoughts?

I thought it would be useful to mention this for potential users to know that Ballerina is capable of providing the above as well.

Author: Replaced

>Be sure to note down the connection and authentication details for connecting to the MySQL server including the hostname, port, username and password.

### Creating a Database and Table

Connect to the MySQL server using the terminal (or any other preferred method) and execute the following commands to 
create a database and table. This tutorial will demonstrate the basic use-case of creating, maintaining and
interacting with a database of employees in an organization.

```roomsql
CREATE DATABASE IF NOT EXISTS Company;

CREATE TABLE Company.Employees (
	employee_id INTEGER AUTO_INCREMENT PRIMARY KEY,
	first_name  VARCHAR(255) NOT NULL,
	last_name   VARCHAR(255) NOT NULL,
	email       VARCHAR(255) NOT NULL,
	phone       VARCHAR(50) NOT NULL ,
	hire_date   DATE NOT NULL,
	manager_id  INTEGER REFERENCES Employees(employee_id),
	job_title   VARCHAR(255) NOT NULL
)
```

### Creating a Ballerina Project

A new Ballerina project can be created by executing the following command in the directory in which you want to create the project.

```shell
bal new data_service
``` 

> >For more information on creating Ballerina packages, see [creating your first Ballerina package](/learn/hello-world/creating-your-first-ballerina-package/#creating-the-package/).

Reviewer 2: This refers to old link. new one I think is /learn/getting-started-with-ballerina

Reviewer 2: @praneesha is this link correct for the new getting-started guide.

Reviewer: @azinneera - What is the new link of this?

Reviewer 3: It should be directed to the new Getting Started guide one Sameera added with https://github.com/ballerina-platform/ballerina-dev-website/pull/3433/

Author: Fixed in learn/getting-started-with-ballerina

    stream<Employee, error?> resultStream = dbClient->query(
        `SELECT * FROM Employees`
    );
    check from Employee employee in resultStream
        do {
            employees.push(employee);
        };
    check resultStream.close();
    return employees;
}

isolated function updateEmployee(Employee emp) returns int|error {
    sql:ExecutionResult result = check dbClient->execute(`
        UPDATE Employees SET
            first_name = ${emp.first_name}, 
            last_name = ${emp.last_name},
            email = ${emp.email},
            phone = ${emp.phone},
            hire_date = ${emp.hire_date}, 
            manager_id = ${emp.manager_id},
            job_title = ${emp.job_title}
        WHERE employee_id = ${emp.employee_id}  
    `);
    int|string? lastInsertId = result.lastInsertId;
    if lastInsertId is int {
        return lastInsertId;
    } else {
        return error("Unable to obtain last insert ID");
    }
}

isolated function removeEmployee(int id) returns int|error {
    sql:ExecutionResult result = check dbClient->execute(`
        DELETE FROM Employees WHERE employee_id = ${id}
    `);
    return result.affectedRowCount;
}
```

> 

Reviewer 2: shall we remove additional new lines

            "year": 2021,
            "month": 12,
            "day": 16
        },
        "manager_id": 1,
        "job_title": "Sales Manager"
    }'
```

Similarly, you can define resource functions within the service for each of our use cases as demonstrated below.

```ballerina
service /employees on new http:Listener(8080) {

    isolated resource function post .(@http:Payload Employee emp) returns int|error? {
        return addEmployee(emp);
    }
    
    isolated resource function get [int id]() returns Employee|error? {
        return getEmployee(id);
    }
    
    isolated resource function get .() returns Employee[]|error? {
        return getAllEmployees();
    }
    
    isolated resource function put .(@http:Payload Employee emp) returns int|error? {
        return updateEmployee(emp);
    }
    
    isolated resource function delete [int id]() returns int|error? {
        return removeEmployee(id);       
    }

}
```

>**Note:** 
>This tutorial does not cover topics such as authentication, rate-limiting, usage monitoring, billing, caching, 
>observability, load balancing and security.
>Ballerina provides features and modules to cover these aspects as well.

> <style> #tree-expand-all , #tree-collapse-all, .cTocElements {display:none;} .cGitButtonContainer {padding-left: 40px;} </style>

Reviewer 2: Do we need this line

Author: Removed
